client/src/BitBakeDocumentSemanticTokensProvider.ts

```diff
@@ -0,0 +1,170 @@
+
+  const tokenModifiersLegend = [
+    'declaration'
+  ]
+  tokenModifiersLegend.forEach((tokenModifier, index) => tokenModifiers.set(tokenModifier, index))
+
+  return new SemanticTokensLegend(tokenTypesLegend, tokenModifiersLegend)
+}
+
+export const legend = generateSemanticTokensLegend()
+
+export class BitBakeDocumentSemanticTokensProvider implements DocumentSemanticTokensProvider {
+  async provideDocumentSemanticTokens (document: TextDocument, token: CancellationToken): Promise<SemanticTokens> {
+    const allTokens = this._parseText(document.getText())
+    const builder = new SemanticTokensBuilder(legend)
+    allTokens.forEach((token) => {
+      builder.push(token.line, token.startCharacter, token.length, this._encodeTokenType(token.tokenType), this._encodeTokenModifiers(token.tokenModifiers))
+    })
+    return builder.build()
+  }
+
+  // Check node_modules/@types/vscode/index.d.ts for more encoding details
+  private _encodeTokenType (tokenType: string): number {
+    if (tokenTypes.has(tokenType)) {
+      return Number(tokenTypes.get(tokenType))
+    } else if (tokenType === 'notInLegend') {
+      return tokenTypes.size + 2
+    }
+    return 0
+  }
+
+  private _encodeTokenModifiers (strTokenModifiers: string[]): number {
+    let result = 0
+    for (let i = 0; i < strTokenModifiers.length; i++) {
+      const tokenModifier = strTokenModifiers[i]
+      if (tokenModifiers.has(tokenModifier)) {
+        result = result | (1 << Number(tokenModifiers.get(tokenModifier)))
+      } else if (tokenModifier === 'notInLegend') {
+        result = result | (1 << tokenModifiers.size + 2)
+      }
+    }
+    return result
+  }
+
+  private _parseText (text: string): IParsedToken[] {
+    /**
+     * This function traverse the document twice. In the first traverse, the function and variable declarations will be matched and stored in an array which will be used to construct another array of regex for the second traverse. In the second traverse, the array of regex is used to match the function and variable references.
+     */
+
+    // TODO: Share tokens with other handlers such as diagnosis provider
+    const resultTokens: IParsedToken[] = []
+    const lines = text.split(/\r\n|\r|\n/)
+    let declaredFunctions: string[] = []
+    let declaredVariables: string[] = []
+
+    const functionStartsWithDefRegex = /(?<!(#.*)|'.*|".*)(?<=def\s+)(?<name>[a-zA-Z0-9_-][\w-]*)(?<operator>:(append|prepend|remove))?(?=\s*\()/
+    const functionStartsWithPythonRegex = /(?<!(#.*)|'.*|".*)(?<=python\s+)(?<name>[a-zA-Z0-9_-][\w-]*)(?<operator>:(append|prepend|remove))?(?=\s*\()/
+    const shellFunctionsRegex = /(?<!(#.*)|'.*|".*)(?<name>[a-zA-Z0-9_-][\w-]*)(?<operator>:(append|prepend|remove))?(?=\s*\(.*\)\s*\{)/
+    const anonymousFunctionsRegex = /(\bpython)(?=\s*\(.*\)\s*\{)/
+    const variableDeclarationRegex = /(?<name>[a-zA-Z0-9_-][\w-]*)(?<operator>:(append|prepend|remove))?(?=\s*=\s*)(?!\s*\()/
```

Comment:
I don't think it takes into account `.=`, `?=`, `+=` and all the family. That being said, I'm not sure how to test it.

Reply:
To see the tokens, you can open the `command palette` and search for the command `Developer: Inspect Editor Tokens and Scopes`. You can just run it and click where you want to see the tokens and scopes.

This Regex only treats a single `=` presence as a declaration. The rest will be treated as references. They still have same coloring applied to them

Reply:
I am not sure "declaration" is the right word here, but thing is it seems they are all the same for that matter. They all could be the first occurrence of the variable, and they could even be the only one. You could do `TEST += 'TEST'` and not having the variable `TEST` used anywhere else in your code.

That being said, I struggle to imagine a test case where it would mean a different behavior for our extension.

So, anyway, I am willing to merge it right now and to fix it only if it causes a real life incorrect behavior.